src/util-lua-sandbox.c

```diff
@@ -0,0 +1,300 @@
+        return NULL;
+    } else {
+        // We can be a bit sloppy on the alloc limit since it's not supposed to be hit.
+        //  ASSERT: ctx->alloc_bytes + nsize > ctx->alloc_bytes
+        if (ctx->alloc_bytes + nsize > ctx->alloc_limit) {
+            // TODO: Trace in a better way
+            return NULL;
+        }
+        void *nptr = SCRealloc(ptr, nsize);
+        if (nptr != NULL) {
+            ctx->alloc_bytes += nsize;
+        }
+        return nptr;
+    }
+}
+
+/*
+** These are the set of libs allowed in the restricted lua sandbox
+*/
+static const luaL_Reg sb_restrictedlibs[] = { { LUA_GNAME, luaopen_base },
+    //  {LUA_LOADLIBNAME, luaopen_package},
+    //  {LUA_COLIBNAME, luaopen_coroutine},
+    { LUA_TABLIBNAME, luaopen_table },
+    //  {LUA_IOLIBNAME, luaopen_io},
+    //  {LUA_OSLIBNAME, luaopen_os},
+    { LUA_STRLIBNAME, luaopen_string }, { LUA_MATHLIBNAME, luaopen_math },
+    { LUA_UTF8LIBNAME, luaopen_utf8 },
+    { LUA_DBLIBNAME, luaopen_sandbox }, // TODO: remove this from restricted
+    { NULL, NULL } };
+
+// TODO: should we block raw* functions?
+// TODO: Will we ever need to block a subset of functions more than one level deep?
+static const sb_block_function sb_restrictedfuncs[] = { { LUA_GNAME, "collectgarbage" },
+    { LUA_GNAME, "dofile" }, { LUA_GNAME, "getmetatable" }, { LUA_GNAME, "loadfile" },
+    { LUA_GNAME, "load" }, { LUA_GNAME, "pcall" }, { LUA_GNAME, "setmetatable" },
+    { LUA_GNAME, "xpcall" },
+
+    { LUA_STRLIBNAME, "rep" }, // TODO:  probably don't need to block this for normal restricted
+                               // since we have memory limit
+    { NULL, NULL } };
+
+static void sb_loadlibs(lua_State *L, const luaL_Reg *libs)
+{
+    const luaL_Reg *lib;
+    /* "require" functions from 'loadedlibs' and set results to global table */
+    for (lib = libs; lib->func; lib++) {
+        luaL_requiref(L, lib->name, lib->func, 1);
+        lua_pop(L, 1); /* remove lib */
+    }
+}
+
+static void sb_block_functions(lua_State *L, const sb_block_function *funcs)
+{
+    const sb_block_function *func;
+
+    // set target functions to nil
+    lua_pushglobaltable(L);
+    for (func = funcs; func->module; func++) {
+        lua_pushstring(L, func->module);
+        lua_gettable(L, -2); // load module to stack
+        lua_pushstring(L, func->name);
+        lua_pushnil(L);
+        lua_settable(L, -3);
+        lua_pop(L, 1); // remove module from the stack
+    }
+    lua_pop(L, 1); // remove global table
+}
+
+LUALIB_API void sb_loadrestricted(lua_State *L)
+{
+    sb_loadlibs(L, sb_restrictedlibs);
+    sb_block_functions(L, sb_restrictedfuncs);
+}
+
+lua_State *sb_newstate(uint64_t alloclimit, uint64_t instructionlimit)
+{
+    sb_lua_state *sb = SCMalloc(sizeof(sb_lua_state));
+    if (sb == NULL) {
+        // Out of memory.  Error code?
+        return NULL;
+    }
+
+    sb->alloc_limit = alloclimit;
+    sb->alloc_bytes = 0;
+    sb->hook_instruction_count = 100;
```

Comment:
what is this and should it be configurable?

Reply:
This appears to be the number of instructions that need to occur before the callback is called to check the instruction count. So Suricata can check the instruction count every 100 instructions here.

I'm not sure if it makes sense to have this configurable. Ideally its as large as possible, while still small enough to hit our granularity of instruction counts.

@J0eJ0h Any downsides to setting it to our maximum instruction count?

Reply:
The main use case would be for memory profiling during execution.  A script might spike allocations really high while running without hitting the hard memory limit but manage to free most of it by the time it completes.  Since profiling is not currently being done, it should be fine since resetinstructioncounter also resets the instruction hook counter.

Per question on the misplaced thread, the allocation limit is enforced by the allocator passed in when the lua state is created.  This lets us capture under the hood allocations for lua in addition to just script-space ones.  The logic is in sb_alloc.

Reply:
So now downsides to just setting it to the maximum.  Allocations will be still limited, but if we wanted more detailed view into the memory allocations over the call into the Lua rule, this is where we would do it?

Reply:
We can do it via the hook periodically, or use the debug apis on the lua state itself, but that's a bit tricky because you don't want to do anything to trigger the allocator inside the allocator.